File: 383.ransom-note.py
#
# @lc app=leetcode id=383 lang=python3
#
# [383] Ransom Note
#

# Time:  O(n)
# Space: O(1)

from collections import defaultdict


# @lc code=start
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        lookup = defaultdict(int)
        for ch in magazine:
            lookup[ch] += 1
        for ch in ransomNote:
            if lookup[ch] == 0:
                return False
            lookup[ch] -= 1
        return True


# @lc code=end

File: 383.ransom-note.go
/*
 * @lc app=leetcode id=383 lang=golang
 *
 * [383] Ransom Note
 */

package leetcode

// @lc code=start
func canConstruct(ransomNote string, magazine string) bool {
	lookup := map[byte]int{}
	for i := 0; i < len(magazine); i++ {
		lookup[magazine[i]]++
	}
	for i := 0; i < len(ransomNote); i++ {
		if lookup[ransomNote[i]] == 0 {
			return false
		}
		lookup[ransomNote[i]]--
	}
	return true
}

// @lc code=end
